File: 02 - build your json parser/python/Lexer.py
class Lexer:
    def __init__(self, text):
        """
        Initialize the lexer with the input text.
        :param text: The input JSON string to be tokenized.
        self._text - The input JSON string.
        self._line - The current line number in the input string.
        self._column - The current column number in the input string.
        self._pos - The current position in the input string.
        self.current_char - The character at the current position.
        """
        self._text = text
        self._line = 1
        self._column = 1
        self._pos = 0
        self.current_char = self._text[self._pos] if self._text else None

    def _next(self):
        """
        Advance the 'pos' pointer and update the 'current_char'.
        If the end of the input string is reached, set 'current_char' to None.
        """
        self._pos += 1
        if self.current_char == '\n':
            self._line += 1
            self._column = 1
        else:
            self._column += 1

        if self._pos >= len(self._text):
            self.current_char = None
        else:
            self.current_char = self._text[self._pos]
            
        

    def _skip_whitespace(self):
        """
        Skip whitespace characters in the input string.
        Advances the 'pos' pointer until a non-whitespace character is found.
        """
        while self.current_char is not None and self.current_char.isspace():
            self._next()


    def _number(self):
        """
        Parse a number from the input string.
        Handles integers, floats, and scientific notation.
        Validates JSON number format (no leading zeros).
        :return: The parsed number as an int or float.
        """
        result = ''
        
        # Handle negative sign
        if self.current_char == '-':
            result += self.current_char
            self._next()
        
        # Parse integer part with leading zero validation
        if self.current_char == '0':
            result += self.current_char
            self._next()
            # After '0', next char must be '.', 'e', 'E', or end of number
            if self.current_char and self.current_char.isdigit():
                self.error()  # Reject leading zeros like 013, 01.5
        elif self.current_char and self.current_char.isdigit():
            # Parse remaining digits (1-9 followed by any digits)
            while self.current_char is not None and self.current_char.isdigit():
                result += self.current_char
                self._next()
        else:
            self.error()  # No digits found

        # Parse decimal part
        if self.current_char == '.':
            result += self.current_char
            self._next()
            while self.current_char is not None and self.current_char.isdigit():
                result += self.current_char
                self._next()

        # Parse scientific notation (e or E)
        if self.current_char and self.current_char.lower() == 'e':
            result += self.current_char
            self._next()
            
            # Handle + or - after e/E
            if self.current_char and self.current_char in '+-':
                result += self.current_char
                self._next()
            
            # Parse exponent digits
            while self.current_char is not None and self.current_char.isdigit():
                result += self.current_char
                self._next()

        # Convert to appropriate type
        if '.' in result or 'e' in result.lower():
            return float(result)
        return int(result)
    
    def _string(self):
        """
        Parse a string from the input string.
        Handles escape sequences including unicode.
        :return: The parsed string without the enclosing quotes.
        """
        result = ''
        self._next()  # Skip the opening quote

        while self.current_char is not None and self.current_char != '"':
            # Check for unescaped control characters (ASCII 0-31)
            # ord is a built-in Python function that returns 
            # the Unicode code point (numeric value) of a character.
            # numeric unicode value less than 32 are control characters
            if ord(self.current_char) < 32:
                self.error()  # Reject unescaped control characters like \n, \t, etc.
            
            if self.current_char == '\\':
                self._next()  # Skip backslash
                if self.current_char == '"':
                    result += '"'
                elif self.current_char == '\\':
                    result += '\\'
                elif self.current_char == '/':
                    result += '/'
                elif self.current_char == 'b':
                    result += '\b'
                elif self.current_char == 'f':
                    result += '\f'
                elif self.current_char == 'n':
                    result += '\n'
                elif self.current_char == 'r':
                    result += '\r'
                elif self.current_char == 't':
                    result += '\t'
                elif self.current_char == 'u':
                    # Handle unicode escape \uXXXX
                    self._next()
                    hex_chars = ''
                    for _ in range(4):
                        if self.current_char and self.current_char in '0123456789abcdefABCDEF':
                            hex_chars += self.current_char
                            self._next()
                        else:
                            self.error()
                    result += chr(int(hex_chars, 16))
                    continue  # Don't advance again
                else:
                    self.error()
            else:
                result += self.current_char
            self._next()

        self._next()  # Skip the closing quote
        return result
    
    def _peek(self):
        """
        Return the _next character without advancing the position.
        :return: The _next character or None if at end of input.
        """
        peek_pos = self._pos + 1
        if peek_pos >= len(self._text):
            return None
        return self._text[peek_pos]

    def _boolean(self):
        """
        Parse a boolean value from the input string.
        :return: The parsed boolean value (True or False).
        """
        if self._text[self._pos:self._pos + 4] == 'true':
            for _ in range(4):
                self._next()
            return 'TRUE'
        elif self._text[self._pos:self._pos + 5] == 'false':
            for _ in range(5):
                self._next()
            return 'FALSE'
        
        
    def _null(self):
        """
        Parse a null value from the input string.
        :return: 'NULL' token type
        """
        if self._text[self._pos:self._pos + 4] == 'null':
            for _ in range(4):
                self._next()
            return 'NULL'
    
        
    def error(self):
        """
        Raise an exception for invalid characters.
        """
        raise Exception(f"Invalid Syntax: at line {self._line} column {self._column} of file.")

    
    def get_next_token(self):
        """
        Tokenize the input string and return the _next token.
        :return: The _next token in the input string.
        """
        while self.current_char is not None:
            if self.current_char.isspace():
                self._skip_whitespace()
                continue

            if self.current_char == '{':
                self._next()
                return ('LBRACE', '{')

            if self.current_char == '}':
                self._next()
                return ('RBRACE', '}')

            if self.current_char == '[':
                self._next()
                return ('LBRACKET', '[')

            if self.current_char == ']':
                self._next()
                return ('RBRACKET', ']')

            if self.current_char == ':':
                self._next()
                return ('COLON', ':')

            if self.current_char == ',':
                self._next()
                return ('COMMA', ',')

            if self.current_char == '"':
                value = self._string()
                return ('STRING', value)

            if self.current_char.isdigit() or (self.current_char == '-' and self._peek() and self._peek().isdigit()):
                value = self._number()
                return ('NUMBER', value)

            if self.current_char == 't' or self.current_char == 'f':
                token_type = self._boolean()
                return (token_type, token_type)

            if self.current_char == 'n':
                token_type = self._null()
                return (token_type, None)

            self.error()

        return ('EOF', None)
    

File: 02 - build your json parser/c++/lexer.cpp
#include "json_parser.hpp"
#include <iostream>
#include <stdexcept>
#include <sstream>
#include <cctype>
#include <iomanip>

// Lexer implementation
Lexer::Lexer(const std::string &input) : text(input), pos(0), line(1), column(1)
{
    currentChar = pos < text.length() ? text[pos] : '\0';
}

void Lexer::advance()
{
    if (currentChar == '\n')
    {
        line++;
        column = 1;
    }
    else
    {
        column++;
    }

    pos++;
    currentChar = pos < text.length() ? text[pos] : '\0';
}

void Lexer::skipWhitespace()
{
    while (currentChar != '\0' && std::isspace(currentChar))
    {
        advance();
    }
}

std::string Lexer::parseString()
{
    std::string result = "";
    advance(); // Skip opening quote

    while (currentChar != '\0' && currentChar != '"')
    {
        if (currentChar == '\\')
        {
            advance(); // Skip backslash
            if (currentChar == '\0')
            {
                error("Unterminated string escape");
            }

            // Handle escape sequences
            switch (currentChar)
            {
            case '"':
                result += '"';
                break;
            case '\\':
                result += '\\';
                break;
            case '/':
                result += '/';
                break;
            case 'b':
                result += '\b';
                break;
            case 'f':
                result += '\f';
                break;
            case 'n':
                result += '\n';
                break;
            case 'r':
                result += '\r';
                break;
            case 't':
                result += '\t';
                break;
            case 'u':            // Unicode escape - simplified
                result += "\\u"; // Keep as-is for simplicity
                break;
            default:
                error("Invalid escape sequence");
            }
        }
        else
        {
            // Check for control characters
            if (currentChar < 32 && currentChar != '\t')
            {
                error("Control character in string");
            }
            result += currentChar;
        }
        advance();
    }

    if (currentChar != '"')
    {
        error("Unterminated string");
    }

    advance(); // Skip closing quote
    return result;
}

std::string Lexer::parseNumber()
{
    std::string result = "";

    // Handle negative sign
    if (currentChar == '-')
    {
        result += currentChar;
        advance();
    }

    // Integer part
    if (currentChar == '0')
    {
        result += currentChar;
        advance();
        // After 0, can't have more digits unless it's a decimal
        if (std::isdigit(currentChar))
        {
            error("Leading zeros not allowed");
        }
    }
    else if (std::isdigit(currentChar))
    {
        while (std::isdigit(currentChar))
        {
            result += currentChar;
            advance();
        }
    }
    else
    {
        error("Invalid number");
    }

    // Decimal part
    if (currentChar == '.')
    {
        result += currentChar;
        advance();

        if (!std::isdigit(currentChar))
        {
            error("Digit expected after decimal point");
        }

        while (std::isdigit(currentChar))
        {
            result += currentChar;
            advance();
        }
    }

    // Exponent part
    if (currentChar == 'e' || currentChar == 'E')
    {
        result += currentChar;
        advance();

        if (currentChar == '+' || currentChar == '-')
        {
            result += currentChar;
            advance();
        }

        if (!std::isdigit(currentChar))
        {
            error("Digit expected in exponent");
        }

        while (std::isdigit(currentChar))
        {
            result += currentChar;
            advance();
        }
    }

    return result;
}

std::string Lexer::parseKeyword()
{
    std::string result = "";

    while (std::isalpha(currentChar))
    {
        result += currentChar;
        advance();
    }

    return result;
}

void Lexer::error(const std::string &msg)
{
    std::ostringstream oss;
    oss << "Lexer error at line " << line << ", column " << column << ": " << msg;
    throw std::runtime_error(oss.str());
}

Token Lexer::getNextToken()
{
    while (currentChar != '\0')
    {
        if (std::isspace(currentChar))
        {
            skipWhitespace();
            continue;
        }

        switch (currentChar)
        {
        case '{':
            advance();
            return Token(TokenType::LBRACE, "{");
        case '}':
            advance();
            return Token(TokenType::RBRACE, "}");
        case '[':
            advance();
            return Token(TokenType::LBRACKET, "[");
        case ']':
            advance();
            return Token(TokenType::RBRACKET, "]");
        case ',':
            advance();
            return Token(TokenType::COMMA, ",");
        case ':':
            advance();
            return Token(TokenType::COLON, ":");
        case '"':
        {
            std::string str = parseString();
            return Token(TokenType::STRING, str);
        }
        case '-':
        case '0':
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
        {
            std::string num = parseNumber();
            return Token(TokenType::NUMBER, num);
        }
        case 't':
        case 'f':
        case 'n':
        {
            std::string keyword = parseKeyword();
            if (keyword == "true")
            {
                return Token(TokenType::TRUE, "true");
            }
            else if (keyword == "false")
            {
                return Token(TokenType::FALSE, "false");
            }
            else if (keyword == "null")
            {
                return Token(TokenType::NULL_TOKEN, "null");
            }
            else
            {
                error("Invalid keyword: " + keyword);
            }
            break;
        }
        default:
            error("Unexpected character: " + std::string(1, currentChar));
        }
    }

    return Token(TokenType::EOF_TOKEN, "");
}